File: my-folder/0036-valid-sudoku/solution.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:

        #for rows
        for i in range(9):
            hm={}
            for j in range(9):
                if board[i][j]=='.':
                    continue
                hm[board[i][j]]=hm.get(board[i][j],0)+1
            for val in hm.values():
                if val>1:
                    return False
        #for columns
        for j in range(9):
            hm={}
            for i in range(9):
                if board[i][j]=='.':
                    continue
                hm[board[i][j]]=hm.get(board[i][j],0)+1
            for val in hm.values():
                if val>1:
                    return False
        #for 3*3 grid
        for row in range(3):
            for col in range(3):
                hm={}
                for i in range(row*3,(row+1)*3):
                    for j in range(col*3,(col+1)*3):
                        if board[i][j]=='.':
                            continue
                        hm[board[i][j]]=hm.get(board[i][j],0)+1
                    for val in hm.values():
                        if val>1:
                            return False
        return True
                
                


        

File: my-folder/0036-valid-sudoku/solution.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer, HashSet<Character>> rowHash = new HashMap<>();
        Map<Integer, HashSet<Character>> colHash = new HashMap<>();
        Map<Integer, HashSet<Character>> subGridHash = new HashMap<>();
        int rows = board.length;
        int cols = board[0].length;
        
        for(int i = 0; i < rows; i++){
            for(int j = 0; j < cols; j++){
                if(board[i][j] == '.') continue;
                
                if(!rowHash.containsKey(i)){
                    rowHash.put(i, new HashSet<Character>());
                }

                if(rowHash.get(i).contains(board[i][j])){
                    return false;
                }

                rowHash.get(i).add(board[i][j]);

                if(!colHash.containsKey(j))
                {
                    colHash.put(j, new HashSet<Character>());
                } 

                if(colHash.get(j).contains(board[i][j])){
                    return false;
                }
                
                colHash.get(j).add(board[i][j]);

                int key = getSubGridKey(i, j);
                if(!subGridHash.containsKey(key))
                {
                    subGridHash.put(key, new HashSet<Character>());
                } 

                if(subGridHash.get(key).contains(board[i][j])){
                    return false;
                }
                
                subGridHash.get(key).add(board[i][j]);
            }
        }

        return true;
    } 
    
    private int getSubGridKey(int row, int col){
        if(row/3 == 0){
            return col/3;
        }
        else if(row/3 == 1){
            return 3 + col/3;
        }
        
        return 6 + col/3;  
    }
}


